pkg/liquidity-mining/test/ChildChainGaugeRegistry.test.ts

```diff
@@ -0,0 +1,135 @@
+      args: [vault.address, mockVE.address, ZERO_ADDRESS],
+    });
+
+    const pseudoMinter = await deploy('L2BalancerPseudoMinter', { args: [vault.address, balToken.address] });
+
+    gaugeImplementation = await deploy('ChildChainGauge', {
+      args: [veDelegationProxy.address, pseudoMinter.address, vault.authorizerAdaptor.address, productVersion],
+    });
+
+    gaugeFactory = await deploy('ChildChainGaugeFactory', {
+      args: [gaugeImplementation.address, factoryVersion, productVersion],
+    });
+
+    otherGaugeFactory = await deploy('ChildChainGaugeFactory', {
+      args: [gaugeImplementation.address, factoryVersion, productVersion],
+    });
+
+    registry = await deploy('ChildChainGaugeRegistry', {
+      args: [vault.address, pseudoMinter.address, gaugeFactory.address],
+    });
+
+    await vault.grantPermissionGlobally(await actionId(pseudoMinter, 'addGaugeFactory'), admin.address);
+    await vault.grantPermissionGlobally(await actionId(registry, 'addGauge'), admin.address);
+    await vault.grantPermissionGlobally(await actionId(registry, 'removeGauge'), admin.address);
+
+    await pseudoMinter.connect(admin).addGaugeFactory(gaugeFactory.address);
+
+    let tx = await gaugeFactory.create(mockBPT.address);
+    let event = expectEvent.inReceipt(await tx.wait(), 'GaugeCreated');
+    gauge = await deployedAt('ChildChainGauge', event.args.gauge);
+
+    tx = await otherGaugeFactory.create(mockBPT.address);
+    event = expectEvent.inReceipt(await tx.wait(), 'GaugeCreated');
+    otherGauge = await deployedAt('ChildChainGauge', event.args.gauge);
+  });
+
+  describe('addGauge', () => {
+    it('can add a valid gauge to the registry', async () => {
+      const tx = await registry.connect(admin).addGauge(gauge.address);
+
+      expectEvent.inReceipt(await tx.wait(), 'GaugeAdded', { gauge: gauge.address });
+      expect(await registry.totalGauges()).to.eq(1);
+      expect((await registry.getGauges(0, 1))[0]).to.eq(gauge.address);
+    });
+
+    it('only privileged account can add a gauge', async () => {
+      await expect(registry.connect(other).addGauge(gauge.address)).to.be.revertedWith('SENDER_NOT_ALLOWED');
+    });
+
+    it('cannot add the same gauge twice', async () => {
+      await registry.connect(admin).addGauge(gauge.address);
+      await expect(registry.connect(admin).addGauge(gauge.address)).to.be.revertedWith('GAUGE_ALREADY_REGISTERED');
+    });
+
+    it('reverts if gauge is not from the factory', async () => {
+      const maliciousGauge = await deploy('MockChildChainGauge', { args: ['test'] });
+      await maliciousGauge.setMockFactory(gaugeFactory.address);
+      await expect(registry.connect(admin).addGauge(maliciousGauge.address)).to.be.revertedWith(
+        'GAUGE_NOT_FROM_FACTORY'
+      );
+    });
+
+    it('reverts if gauge factory is invalid', async () => {
+      await expect(registry.connect(admin).addGauge(otherGauge.address)).to.be.revertedWith('INVALID_GAUGE_FACTORY');
+    });
```

Comment:
These look good :+1: 

Following the style of other unit tests, you could also use contexts to describe what's happening, and rename the `it` blocks to just `it('reverts')`.

E.g.:

```typescript
context('when the caller is not authorized', () => {
  it('reverts', async () => {
    await expect(registry.connect(other).addGauge(gauge.address)).to.be.revertedWith('SENDER_NOT_ALLOWED');
  });
});
```

And so on and so forth for the rest of the invalid cases.

For the valid one:

```typescript
context('when the gauge, factory and caller are valid', () => {
  itAddsGaugeCorrectly();
});
```

Reply:
```javascript
context('when the gauge, factory and caller are valid', () => {
  itAddsGaugeCorrectly();
});
```
What is the rationale behind not inlining this? Because if you have to jump to the itAddsGaugeCorrectly implementation, you can't see the context description anymore. I understand it is useful when you have to call itAddsGaugeCorrectly in multiple contexts, but it should be inlined in the case it is used only once. Right?

Reply:
I agree that it's a tradeoff, and for this test in particular it's a gray area.

> This modification introduces two extra lines to the code and adds an additional nesting level to the CLI output. While it may be reasonable to implement this when handling multiple reverts within a single context, the rationale for applying it to a single revert is unclear. I am interested in hearing any arguments supporting this approach.

In general we use `context` for more complex tests, and I agree that it's not necessarily justified for these tests. I think the 'pros' of nesting is that it follows the style of other tests, it is more maintainable in some sense, and personally I think the outputs look slightly better. Using contexts also helps me reason about covering all the cases that can go wrong, but again that's a personal appreciation.
None of these reasons are very strong for these tests in particular; as you say the readability gets worse by nesting, and just inlining is more straightforward.

> What is the rationale behind not inlining this? Because if you have to jump to the itAddsGaugeCorrectly implementation, you can't see the context description anymore. I understand it is useful when you have to call itAddsGaugeCorrectly in multiple contexts, but it should be inlined in the case it is used only once. Right?

Same as above; maybe I'm too used to reading other tests that follow that style. Again for this case which is relatively simple maybe it doesn't add much.

All in all, I think the initial implementation is also just fine.